sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,239 @@
+    >>> iris = datasets.load_iris()
+    >>> rng = np.random.RandomState(42)
+    >>> random_unlabeled_points = rng.rand(len(iris.target)) < 0.3
+    >>> labels = np.copy(iris.target)
+    >>> labels[random_unlabeled_points] = -1
+    >>> self_training_model.fit(iris.data, labels)
+    ... # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    SelfTrainingClassifier(...)
+
+    References
+    ----------
+    David Yarowsky. 1995. Unsupervised word sense disambiguation rivaling
+    supervised methods. In Proceedings of the 33rd annual meeting on
+    Association for Computational Linguistics (ACL '95). Association for
+    Computational Linguistics, Stroudsburg, PA, USA, 189-196. DOI:
+    https://doi.org/10.3115/981658.981684
+    """
+    def __init__(self,
+                 base_estimator,
+                 threshold=0.75,
+                 max_iter=100):
+        self.base_estimator = base_estimator
+        self.threshold = threshold
+        self.max_iter = max_iter
+
+    def fit(self, X, y):
+        """
+        Fits this ``SelfTrainingClassifier`` to dataset, using the
+        base_estimator passed. The fitted base_estimator is stored as
+        ``self.base_estimator_``.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        y : array-like, shape = (n_samples, 1)
+            array representing the labels
+
+        Returns
+        -------
+        self: returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+        _check_estimator(self.base_estimator)
+        self.base_estimator_ = clone(self.base_estimator)
+
+        if not 0 <= self.max_iter:
+            raise ValueError("max_iter must be >= 0")
+
+        if not 0 <= self.threshold < 1:
+            raise ValueError("threshold must be in [0,1)")
+
+        # Data usable for supervised training
+        self.X_labeled_ = X[safe_mask(X, np.where(y != -1))][0]
+        self.y_labeled_ = y[safe_mask(y, np.where(y != -1))][0]
+        self.y_labeled_iter_ = np.full_like(self.y_labeled_, 0)
+
+        # Unlabeled data
+        X_unlabeled = X[safe_mask(X, np.where(y == -1))][0]
+        y_unlabeled = y[safe_mask(y, np.where(y == -1))][0]
+
+        iter = 0
+        while len(self.X_labeled_) < len(X) and iter < self.max_iter:
+            iter += 1
+            self.base_estimator_.fit(self.X_labeled_, self.y_labeled_)
+
+            # Select predictions where confidence is above the threshold
+            predict_proba = self.base_estimator_.predict_proba(X_unlabeled)
+
+            pred = np.argmax(predict_proba, axis=1)
+            max_proba = np.max(predict_proba, axis=1)
+
+            confident = np.where(max_proba > self.threshold)[0]
+
+            # Add newly labeled confident predictions to the dataset
+            self.X_labeled_ = np.append(
+                self.X_labeled_, X_unlabeled[confident], axis=0)
```

Comment:
our preferred convention would be for you to split the line after the comma rather than after the parenthesis:
```py
            self.X_labeled_ = np.append(self.X_labeled_,
                                        X_unlabeled[confident], axis=0)
```